node/core/runtime-api/src/lib.rs

```diff
@@ -355,33 +355,55 @@ where
 			let sender = $sender;
 			let version: u32 = $version;	// enforce type for the version expression
 			let runtime_version = client.api_version_parachain_host(relay_parent).await
-				.unwrap_or_else(|e| {
-					gum::warn!(
-						target: LOG_TARGET,
-						"cannot query the runtime API version: {}",
-						e,
-					);
-					Some(0)
+				.map_err(|e| {
+					if let ApiError::UnknownBlock(s) = e {
+						// We want a specific indication for this error. It generally indicates that the runtime api is
+						// called for a block which is already pruned. Legitimate reasons for this may be block restart.
+						// Separate value in `RuntimeApiError` enum allows the caller to handle it properly.
+						gum::warn!(
+							target: LOG_TARGET,
+							"cannot query the runtime API version due to UnknownBlock error: {}",
+							s,
+						);
+						RuntimeApiError::UnknownBlock{runtime_api_name: stringify!($api_name)}
+					} else {
+						gum::warn!(
+							target: LOG_TARGET,
+							"cannot query the runtime API version due to error: {}",
+							e,
+						);
+						RuntimeApiError::Execution{runtime_api_name: stringify!($api_name), source: std::sync::Arc::new(e)}
+					}
 				})
-				.unwrap_or_else(|| {
-					gum::warn!(
-						target: LOG_TARGET,
-						"no runtime version is reported"
-					);
-					0
+				.map(|v| {
+					match v {
+						Some(v) => v,
+						None => {
+							gum::warn!(
+								target: LOG_TARGET,
+								"no runtime version is reported"
+							);
+							0
+						}
+					}
 				});
 
-			let res = if runtime_version >= version {
-				client.$api_name(relay_parent $(, $param.clone() )*).await
+			let res = match runtime_version {
+				Ok(runtime_version) if runtime_version >= version => {
+					client.$api_name(relay_parent $(, $param.clone() )*).await
```

Comment:
Mind explaining what's going on here? I'm not familiar with the syntax being used. Particularly the $ and * uses.

Reply:
Yes.... the code is part of a macro. Check the declaration [here](https://github.com/paritytech/polkadot/pull/6288/files#diff-1c816593e067f095a3388e13824a0729ac4d1d9b7fcf96ab2ea6761ba423a9dcR353)

It's a huge topic, but in nutshell - this is a declarative macro. Think about it as macros in C but  on heavy steroids. The syntax is close to pattern matching - you can call the macro with different parameters. The variables with $ are the matched macro parameters. The parameters are [here](https://github.com/paritytech/polkadot/pull/6288/files#diff-1c816593e067f095a3388e13824a0729ac4d1d9b7fcf96ab2ea6761ba423a9dcR354).

`($($param:expr),*)` on line 354 means 'multiple expressions separated by comma' while `$(, $param.clone() )*` means 'put all the expressions in $param here and separate them with comma'. 

Really quick and lame explanation but it's a huge topic.